Snakefile

```diff
@@ -0,0 +1,402 @@
+import subprocess
+
+#################################################################
+####################### general setup ###########################
+#################################################################
+
+# certain files can be fetched from the database or downloaded from our S3 bucket.
+# config["fetch_from_database"] dictates this
+# We could implement this via remote snakemake inputs (such as in nextstrain/ncov),
+# conditional rules (a bit of an anti-pattern) or ruleorders.
+ruleorder: download_previous_biosample > fetch_biosample_from_database
+ruleorder: download_previous_raw_ndjson > fetch_from_database
+if config.get("fetch_from_database", False):
+    ruleorder: fetch_biosample_from_database > download_previous_biosample
+    ruleorder: fetch_from_database > download_previous_raw_ndjson
+
+send_notifications = "SLACK_CHANNELS" in os.environ and "SLACK_TOKEN" in os.environ
+
+#################################################################
+################ work out what steps to run #####################
+#################################################################
+
+def _compute_targets(wildcards):
+    database = config["database_name"]    
+    targets = [f"data/{database}/upload.done"]
+    if config.get("trigger_preprocessing", False):
+        targets.append(f"data/{database}/trigger-preprocessing.done")
+    if send_notifications:
+        targets.append(f"data/{database}/notify.done")
+    if config.get("fetch_from_database", False):
+        targets.append(f"data/{database}/raw.upload.done")
+    return targets
+
+rule all:
+    input: _compute_targets
+
+#################################################################
+###################### rule definitions #########################
+#################################################################
+
+
+def run_shell_command_n_times(cmd, msg, cleanup_failed_cmd, retry_num=5):
+    attempt = 0
+    while attempt < retry_num:
+        print(f"{msg} attempt number {attempt}")
+        try:
+            subprocess.run(cmd, shell=True, check=True)
+            break
+        except subprocess.CalledProcessError:
+            print("...FAILED")
+            attempt+=1
+            shell("{cleanup_failed_cmd} && sleep 10")
+    else:
+        print(msg + f" has FAILED {retry_num} times. Exiting.")
+        raise Exception("function run_shell_command_n_times has failed")
```

Comment:
Post-merge change: I see that this approach existed in the original shell script, but now that we are using Snakemake, we can use [its `--restart-times` argument](https://snakemake.readthedocs.io/en/stable/executing/cli.html) to specify how many times we want to retry any failing rule. This logic will apply to all rules in the workflow, but that should be OK. Then we can drop this kind of custom function that tries to do the same thing inside `run` blocks and just make the shell call in a normal `shell` block.

Reply:
> This logic will apply to all rules in the workflow, but that should be OK. 

I'm not so sure about this. There are a few very long-running rules in this workflow, which means if they fail late in a deterministic way, we'd spend hours retrying them only to see the same late failures again, before reporting failure of the whole workflow.

Reply:
Yeah, that's a fair point. I wrote that comment just before asking about the NDJSON download times, etc. Probably better to wait for Snakemake to support this as a rule-specific option, as you mention.